runtime/pkg/rduckdb/db.go

```diff
@@ -0,0 +1,1068 @@
+	d.logger.Debug("drop table", slog.String("name", name))
+	err := d.writeSem.Acquire(ctx, 1)
+	if err != nil {
+		return err
+	}
+	defer d.writeSem.Release(1)
+
+	// pull latest changes from remote
+	err = d.pullFromRemote(ctx)
+	if err != nil {
+		return fmt.Errorf("drop: unable to pull from remote: %w", err)
+	}
+
+	// check if table exists
+	meta, err := d.tableMeta(name)
+	if err != nil {
+		if errors.Is(err, errNotFound) {
+			return fmt.Errorf("drop: Table %q not found", name)
+		}
+		return fmt.Errorf("drop: unable to get table meta: %w", err)
+	}
+
+	// drop the table from remote
+	d.localDirty = true
+	err = d.deleteRemote(ctx, name, "")
+	if err != nil {
+		return fmt.Errorf("drop: unable to drop table %q from remote: %w", name, err)
+	}
+	// no errors after this point since background goroutine will eventually sync the local db
+
+	// mark table as deleted in local
+	meta.Deleted = true
+	err = d.writeTableMeta(name, meta)
+	if err != nil {
+		d.logger.Debug("drop: error in writing table meta", slog.String("name", name), slog.String("error", err.Error()))
+		return nil
+	}
+
+	// reopen db handle
+	err = d.reopen(ctx)
+	if err != nil {
+		d.logger.Debug("drop: error in reopening db", slog.String("error", err.Error()))
+		return nil
+	}
+	d.localDirty = false
+	return nil
+}
+
+func (d *db) RenameTable(ctx context.Context, oldName, newName string) error {
+	d.logger.Debug("rename table", slog.String("from", oldName), slog.String("to", newName))
+	if strings.EqualFold(oldName, newName) {
+		return fmt.Errorf("rename: Table with name %q already exists", newName)
+	}
+	err := d.writeSem.Acquire(ctx, 1)
+	if err != nil {
+		return err
+	}
+	defer d.writeSem.Release(1)
+
+	// pull latest changes from remote
+	err = d.pullFromRemote(ctx)
+	if err != nil {
+		return fmt.Errorf("rename: unable to pull from remote: %w", err)
+	}
+
+	oldMeta, err := d.tableMeta(oldName)
+	if err != nil {
+		if errors.Is(err, errNotFound) {
+			return fmt.Errorf("rename: Table %q not found", oldName)
+		}
+		return fmt.Errorf("rename: unable to get table meta: %w", err)
+	}
+
+	// copy the old table to new table
+	newVersion := newVersion()
+	err = copyDir(filepath.Join(d.localPath, newName, newVersion), filepath.Join(d.localPath, oldName, oldMeta.Version))
+	if err != nil {
+		return fmt.Errorf("rename: copy table failed: %w", err)
+	}
+
+	// rename the underlying table
+	err = renameTable(ctx, filepath.Join(d.localPath, newName, newVersion, "data.db"), oldName, newName)
+	if err != nil {
+		return fmt.Errorf("rename: rename table failed: %w", err)
+	}
+
+	d.localDirty = true
+	// sync the new table and new version
+	meta := &tableMeta{
+		Name:           newName,
+		Version:        newVersion,
+		CreatedVersion: newVersion,
+		Type:           oldMeta.Type,
+		SQL:            oldMeta.SQL,
+	}
+	if err := d.pushToRemote(ctx, newName, oldMeta, meta); err != nil {
+		return fmt.Errorf("rename: unable to replicate new table: %w", err)
+	}
+
+	// TODO :: fix this
```

Comment:
We need to think solutions for this issue. Few solutions that are on top of my mind:
1. Use a single metadata file for entire database so that all operations are atomically done but need to think about concurrent write contentions.
2. Use the `meta.json` to store `renamed_from_table` and `renamed_from_table_version` so that any table with the given version can be skipped even if delete fails. 

This is an important fix but it should be okay to fix this later IMO. As of now most renames are from staging table to main table and table are always created with replace so it should be okay if staging table is not properly deleted during renames.

Reply:
It would be nice to avoid a single metadata file for the entire database – it would make concurrent writes trickier and add complexity because the file structure is no longer sufficient metadata.

I agree this doesn't need to be solved now (worst case is some garbage old tables that are not used). I'm not sure if I understand your option 2 correctly, maybe this is what you mean, but IMO the way to solve this problem would be:
1. Before rename, update the old table's `meta.json` to have `renaming_to_table: new_name` and `renaming_to_version: new_version`.
2. Copy the table to the `new_name` and `new_version`
3. Delete the old table. If this fails, during a later garbage collection (or "vacuum" or whatever you want to call it), if we see a table with `renaming_to_table` where that table and version actually exists, we can garbage collect the old table.

Reply:
Yeah something similar. I meant I would add `renamed_from_table` and `renamed_from_table_version` to the new `meta.json` and if we encounter a case where that table and version exists we skip that table and version and garbage collect that. 
So essentially it is keeping the track at new table vs old table. 
Both are fine IMO. Depending upon where we keep it we need to make sure that tables are processed in that order(either `created_version` asc or `created_version` desc).